packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -446,5 +452,685 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
+    if (firstChild == null) {
+      performResize();
+      return;
+    }
+
+    // Layout slotted children.
+    _backButton.layout(constraints.loosen(), parentUsesSize: true);
+    _nextButton.layout(constraints.loosen(), parentUsesSize: true);
+    _nextButtonDisabled.layout(constraints.loosen(), parentUsesSize: true);
+
+    final double subsequentPageButtonsWidth =
+      _backButton.size.width + _nextButton.size.width;
+    double currentButtonPosition = 0.0;
+    double toolbarWidth; // The width of the whole widget.
+    double firstPageWidth;
+    int currentPage = 0;
+    int i = -1;
+    visitChildren((RenderObject renderObjectChild) {
+      i++;
+      final RenderBox child = renderObjectChild as RenderBox;
+      final ToolbarItemsParentData childParentData =
+          child.parentData as ToolbarItemsParentData;
+      childParentData.shouldPaint = false;
+
+      // Skip slotted children and children on pages after the visible page.
+      if (childToSlot.containsKey(child) || currentPage > _page) {
+        return;
+      }
+
+      double paginationButtonsWidth = 0.0;
+      if (currentPage == 0) {
+        // If this is the last child, it's ok to fit without a forward button.
+        paginationButtonsWidth =
+            i == childCount - 1 ? 0.0 : _nextButton.size.width;
+      } else {
+        paginationButtonsWidth = subsequentPageButtonsWidth;
+      }
+
+      // The width of the menu is set by the first page.
+      child.layout(
+        BoxConstraints.loose(Size(
+          (currentPage == 0 ? constraints.maxWidth : firstPageWidth) - paginationButtonsWidth,
+          constraints.maxHeight,
+        )),
+        parentUsesSize: true,
+      );
+
+      // If this child causes the current page to overflow, move to the next
+      // page and relayout the child.
+      final double currentWidth =
+          currentButtonPosition + paginationButtonsWidth + child.size.width;
+      if (currentWidth > constraints.maxWidth) {
+        currentPage++;
+        currentButtonPosition = _backButton.size.width + dividerWidth;
+        paginationButtonsWidth = _backButton.size.width + _nextButton.size.width;
+        child.layout(
+          BoxConstraints.loose(Size(
+            firstPageWidth - paginationButtonsWidth,
+            constraints.maxHeight,
+          )),
+          parentUsesSize: true,
+        );
+      }
+      childParentData.offset = Offset(currentButtonPosition, 0.0);
+      currentButtonPosition += child.size.width + dividerWidth;
+      childParentData.shouldPaint = currentPage == page;
+
+      if (currentPage == 0) {
+        firstPageWidth = currentButtonPosition + _nextButton.size.width;
+      }
+      if (currentPage == page) {
+        toolbarWidth = currentButtonPosition;
+      }
+    });
+
+    // It shouldn't be possible to navigate beyond the last page.
+    assert(page <= currentPage);
+
+    // Position page nav buttons.
+    final ToolbarItemsParentData nextButtonParentData =
```

Comment:
nit: these three vars could all be declared inside the `if (currentPage > 0) {`, no?

Reply:
Good catch 👍